agent/runner.go

```diff
@@ -62,7 +66,15 @@ func (r *Runner) Run(runnerCtx context.Context) error {
 		return err
 	}
 	if work == nil {
-		return nil
+		return ErrNoWorkflow
+	}
+
+	// if ephemeral, taint the agent before running any workload.
+	if r.ephemeral {
+		err = r.client.TaintAgent(runnerCtx)
+		if err != nil {
+			return fmt.Errorf("tainting agent: %w", err)
+		}
```

Comment:
It is possible to move out, but the ordering is important here.

The agent needs to be tainted after receiving the workflow, just before it runs it.

- If it is tainted _before_, then it will never receive one.
- If it is tainted _after_ a workflow, and that workflow causes the agent to restart in some way, it will never be tainted.

I'm open to alternative ways we can achieve this. It could be handled server side instead, but I was trying as much as possible to limit the scope of the changes.

Reply:
This is only probable with a malicious workflow. If there was a workflow using privileged steps or was running on a local backend agent, it could potentially take over the agent before it is "tainted", and pull whatever workflows it wants until it gets some information or run some other nefarious action.

My goal with this work is to enable me to run privileged steps on my agents by making them ephemeral. This is the first step in doing that.